File: automation-architect/core/analysis/security_scanner.py
"""
Security Scanner - 安全漏洞掃描器
檢測常見安全漏洞和風險
"""

import re
from dataclasses import dataclass
from pathlib import Path
from typing import Dict, List, Optional

try:
    from loguru import logger
except ImportError:
    import logging
    logger = logging.getLogger(__name__)


@dataclass
class SecurityIssue:
    """安全問題數據類"""
    type: str
    severity: str  # critical, high, medium, low
    message: str
    file: str
    line: int
    cwe_id: Optional[str] = None
    recommendation: Optional[str] = None


class SecurityScanner:
    """
    安全漏洞掃描器
    
    檢測常見安全問題：
    - 硬編碼密鑰和敏感信息
    - SQL 注入風險
    - XSS 漏洞
    - 不安全的加密算法
    - 路徑遍歷漏洞
    - CSRF 風險
    """
    
    def __init__(self, config: Optional[Dict] = None):
        """
        初始化安全掃描器
        
        Args:
            config: 掃描配置
        """
        self.config = config or {}
        self._init_patterns()
        logger.info('SecurityScanner initialized')
    
    def _init_patterns(self):
        """初始化安全檢測模式"""
        # 硬編碼密鑰模式
        self.secret_patterns = [
            (r'password\s*=\s*["\'][^"\']+["\']', 'hardcoded-password', 'Hardcoded password detected'),
            (r'api[_-]?key\s*=\s*["\'][^"\']+["\']', 'hardcoded-api-key', 'Hardcoded API key detected'),
            (r'secret\s*=\s*["\'][^"\']+["\']', 'hardcoded-secret', 'Hardcoded secret detected'),
            (r'token\s*=\s*["\'][^"\']+["\']', 'hardcoded-token', 'Hardcoded token detected'),
            (r'private[_-]?key\s*=\s*["\'][^"\']+["\']', 'hardcoded-private-key', 'Hardcoded private key detected'),
        ]
        
        # SQL 注入模式
        self.sql_injection_patterns = [
            (r'execute\s*\(\s*["\'].*?\+.*?["\']', 'sql-injection', 'Potential SQL injection vulnerability'),
            (r'query\s*\(\s*["\'].*?%s.*?["\']', 'sql-injection', 'Potential SQL injection vulnerability'),
            (r'raw\s*\(\s*["\'].*?\+.*?["\']', 'sql-injection', 'Potential SQL injection vulnerability'),
        ]
        
        # XSS 模式
        self.xss_patterns = [
            (r'innerHTML\s*=\s*', 'xss', 'Potential XSS vulnerability via innerHTML'),
            (r'document\.write\s*\(', 'xss', 'Potential XSS vulnerability via document.write'),
            (r'eval\s*\(', 'eval-usage', 'Use of eval() can lead to code injection'),
        ]
        
        # 不安全的加密
        self.crypto_patterns = [
            (r'MD5\s*\(', 'weak-crypto', 'MD5 is cryptographically broken'),
            (r'SHA1\s*\(', 'weak-crypto', 'SHA1 is cryptographically weak'),
            (r'DES\s*\(', 'weak-crypto', 'DES is insecure'),
        ]
        
        # 路徑遍歷
        self.path_traversal_patterns = [
            (r'open\s*\([^)]*\.\./[^)]*\)', 'path-traversal', 'Potential path traversal vulnerability'),
            (r'file\s*\([^)]*\.\./[^)]*\)', 'path-traversal', 'Potential path traversal vulnerability'),
        ]
    
    async def scan(
        self,
        code_path: str,
        severity_filter: Optional[List[str]] = None
    ) -> List[SecurityIssue]:
        """
        執行安全掃描
        
        Args:
            code_path: 代碼文件或目錄路徑
            severity_filter: 嚴重程度過濾 (可選)
            
        Returns:
            List[SecurityIssue]: 安全問題列表
        """
        logger.info(f'Starting security scan for: {code_path}')
        
        path = Path(code_path)
        issues = []
        
        if path.is_file():
            issues.extend(await self._scan_file(path))
        elif path.is_dir():
            for file_path in self._get_scannable_files(path):
                issues.extend(await self._scan_file(file_path))
        
        # 過濾嚴重程度
        if severity_filter:
            issues = [i for i in issues if i.severity in severity_filter]
        
        logger.info(f'Security scan completed. Found {len(issues)} issues')
        
        return issues
    
    async def _scan_file(self, file_path: Path) -> List[SecurityIssue]:
        """
        掃描單個文件
        
        Args:
            file_path: 文件路徑
            
        Returns:
            List[SecurityIssue]: 安全問題列表
        """
        issues = []
        
        try:
            with open(file_path, 'r', encoding='utf-8') as f:
                lines = f.readlines()
            
            for line_num, line in enumerate(lines, 1):
                # 檢查硬編碼密鑰
                issues.extend(self._check_secrets(line, file_path, line_num))
                
                # 檢查 SQL 注入
                issues.extend(self._check_sql_injection(line, file_path, line_num))
                
                # 檢查 XSS
                issues.extend(self._check_xss(line, file_path, line_num))
                
                # 檢查不安全的加密
                issues.extend(self._check_crypto(line, file_path, line_num))
                
                # 檢查路徑遍歷
                issues.extend(self._check_path_traversal(line, file_path, line_num))
        
        except Exception as e:
            logger.error(f'Error scanning file {file_path}: {e}')
        
        return issues
    
    def _check_secrets(
        self,
        line: str,
        file_path: Path,
        line_num: int
    ) -> List[SecurityIssue]:
        """檢查硬編碼密鑰"""
        issues = []
        
        for pattern, issue_type, message in self.secret_patterns:
            if re.search(pattern, line, re.IGNORECASE):
                issues.append(SecurityIssue(
                    type=issue_type,
                    severity='critical',
                    message=message,
                    file=str(file_path),
                    line=line_num,
                    cwe_id='CWE-798',
                    recommendation='Use environment variables or secure vaults for sensitive data'
                ))
        
        return issues
    
    def _check_sql_injection(
        self,
        line: str,
        file_path: Path,
        line_num: int
    ) -> List[SecurityIssue]:
        """檢查 SQL 注入風險"""
        issues = []
        
        for pattern, issue_type, message in self.sql_injection_patterns:
            if re.search(pattern, line, re.IGNORECASE):
                issues.append(SecurityIssue(
                    type=issue_type,
                    severity='high',
                    message=message,
                    file=str(file_path),
                    line=line_num,
                    cwe_id='CWE-89',
                    recommendation='Use parameterized queries or ORM to prevent SQL injection'
                ))
        
        return issues
    
    def _check_xss(
        self,
        line: str,
        file_path: Path,
        line_num: int
    ) -> List[SecurityIssue]:
        """檢查 XSS 風險"""
        issues = []
        
        for pattern, issue_type, message in self.xss_patterns:
            if re.search(pattern, line, re.IGNORECASE):
                severity = 'critical' if 'eval' in issue_type else 'high'
                cwe = 'CWE-95' if 'eval' in issue_type else 'CWE-79'
                
                issues.append(SecurityIssue(
                    type=issue_type,
                    severity=severity,
                    message=message,
                    file=str(file_path),
                    line=line_num,
                    cwe_id=cwe,
                    recommendation='Sanitize user input and use safe DOM manipulation methods'
                ))
        
        return issues
    
    def _check_crypto(
        self,
        line: str,
        file_path: Path,
        line_num: int
    ) -> List[SecurityIssue]:
        """檢查不安全的加密算法"""
        issues = []
        
        for pattern, issue_type, message in self.crypto_patterns:
            if re.search(pattern, line, re.IGNORECASE):
                issues.append(SecurityIssue(
                    type=issue_type,
                    severity='high',
                    message=message,
                    file=str(file_path),
                    line=line_num,
                    cwe_id='CWE-327',
                    recommendation='Use SHA-256 or stronger cryptographic algorithms'
                ))
        
        return issues
    
    def _check_path_traversal(
        self,
        line: str,
        file_path: Path,
        line_num: int
    ) -> List[SecurityIssue]:
        """檢查路徑遍歷風險"""
        issues = []
        
        for pattern, issue_type, message in self.path_traversal_patterns:
            if re.search(pattern, line):
                issues.append(SecurityIssue(
                    type=issue_type,
                    severity='high',
                    message=message,
                    file=str(file_path),
                    line=line_num,
                    cwe_id='CWE-22',
                    recommendation='Validate and sanitize file paths, use whitelisting'
                ))
        
        return issues
    
    def _get_scannable_files(self, directory: Path) -> List[Path]:
        """獲取可掃描的文件"""
        scannable_extensions = {
            '.py', '.js', '.ts', '.java', '.go', '.rs',
            '.cpp', '.c', '.php', '.rb', '.cs'
        }
        
        files = []
        import os
        for root, _, filenames in os.walk(directory):
            for filename in filenames:
                file_path = Path(root) / filename
                if file_path.suffix in scannable_extensions:
                    files.append(file_path)
        
        return files
    
    def generate_report(self, issues: List[SecurityIssue]) -> Dict:
        """
        生成安全掃描報告
        
        Args:
            issues: 安全問題列表
            
        Returns:
            Dict: 掃描報告
        """
        severity_counts = {
            'critical': 0,
            'high': 0,
            'medium': 0,
            'low': 0
        }
        
        for issue in issues:
            severity_counts[issue.severity] = severity_counts.get(issue.severity, 0) + 1
        
        return {
            'total_issues': len(issues),
            'severity_counts': severity_counts,
            'issues_by_type': self._group_by_type(issues),
            'critical_files': self._get_critical_files(issues),
            'recommendations': self._get_recommendations(issues)
        }
    
    def _group_by_type(self, issues: List[SecurityIssue]) -> Dict[str, int]:
        """按類型分組問題"""
        type_counts = {}
        for issue in issues:
            type_counts[issue.type] = type_counts.get(issue.type, 0) + 1
        return type_counts
    
    def _get_critical_files(self, issues: List[SecurityIssue]) -> List[str]:
        """獲取包含嚴重問題的文件"""
        critical_files = set()
        for issue in issues:
            if issue.severity == 'critical':
                critical_files.add(issue.file)
        return list(critical_files)
    
    def _get_recommendations(self, issues: List[SecurityIssue]) -> List[str]:
        """獲取修復建議"""
        recommendations = set()
        for issue in issues:
            if issue.recommendation:
                recommendations.add(issue.recommendation)
        return list(recommendations)


# 便利函數
async def scan_for_security_issues(
    code_path: str,
    config: Optional[Dict] = None
) -> List[SecurityIssue]:
    """
    便利函數：執行安全掃描
    
    Args:
        code_path: 代碼路徑
        config: 配置
        
    Returns:
        List[SecurityIssue]: 安全問題列表
    """
    scanner = SecurityScanner(config)
    return await scanner.scan(code_path)

File: mcp-servers/security-scanner.js
#!/usr/bin/env node

/**
 * Security Scanner MCP Server
 * Enterprise-grade security scanning for vulnerabilities, dependencies, and secrets
 * 
 * @module security-scanner
 * @author SLASolve Team
 * @license MIT
 */

import { Server } from '@modelcontextprotocol/sdk/server/index.js';
import { StdioServerTransport } from '@modelcontextprotocol/sdk/server/stdio.js';
import {
  CallToolRequestSchema,
  ListToolsRequestSchema,
  ErrorCode,
  McpError
} from '@modelcontextprotocol/sdk/types.js';
import { z } from 'zod';

// Validation schemas
const ScanVulnerabilitiesSchema = z.object({
  code: z.string().min(1, 'Code content is required'),
  language: z.string().optional().default('javascript'),
  severity: z.enum(['critical', 'high', 'medium', 'low', 'all']).optional().default('all')
});

const CheckDependenciesSchema = z.object({
  dependencies: z.record(z.string()).or(z.array(z.object({
    name: z.string(),
    version: z.string()
  }))),
  ecosystem: z.string().optional().default('npm')
});

const AnalyzeSecretsSchema = z.object({
  content: z.string().min(1, 'Content is required'),
  strictMode: z.boolean().optional().default(true)
});

/**
 * Security scanning engine
 */
class SecurityScanner {
  constructor() {
    this.vulnerabilityPatterns = this._initVulnerabilityPatterns();
    this.secretPatterns = this._initSecretPatterns();
  }

  /**
   * Scan code vulnerabilities
   */
  scanVulnerabilities(code, language = 'javascript', severity = 'all') {
    const scan = {
      timestamp: new Date().toISOString(),
      language,
      severity,
      vulnerabilities: [],
      summary: {
        total: 0,
        critical: 0,
        high: 0,
        medium: 0,
        low: 0
      }
    };

    // Scan for SQL injection
    const sqlInjection = this._detectSQLInjection(code);
    scan.vulnerabilities.push(...sqlInjection);

    // Scan for XSS
    const xss = this._detectXSS(code);
    scan.vulnerabilities.push(...xss);

    // Scan for insecure crypto
    const crypto = this._detectInsecureCrypto(code);
    scan.vulnerabilities.push(...crypto);

    // Scan for command injection
    const cmdInjection = this._detectCommandInjection(code);
    scan.vulnerabilities.push(...cmdInjection);

    // Scan for path traversal
    const pathTraversal = this._detectPathTraversal(code);
    scan.vulnerabilities.push(...pathTraversal);

    // Scan for insecure deserialization
    const deserialization = this._detectInsecureDeserialization(code);
    scan.vulnerabilities.push(...deserialization);

    // Filter by severity
    if (severity !== 'all') {
      scan.vulnerabilities = scan.vulnerabilities.filter(v => v.severity === severity);
    }

    // Calculate summary
    for (const vuln of scan.vulnerabilities) {
      scan.summary.total++;
      scan.summary[vuln.severity]++;
    }

    return scan;
  }

  /**
   * Check dependency security
   */
  checkDependencies(dependencies, ecosystem = 'npm') {
    const check = {
      timestamp: new Date().toISOString(),
      ecosystem,
      totalDependencies: 0,
      vulnerableDependencies: [],
      summary: {
        vulnerable: 0,
        critical: 0,
        high: 0,
        medium: 0,
        low: 0
      },
      recommendations: []
    };

    // Normalize dependencies format
    const deps = Array.isArray(dependencies) 
      ? dependencies 
      : Object.entries(dependencies).map(([name, version]) => ({ name, version }));

    check.totalDependencies = deps.length;

    // Check each dependency
    for (const dep of deps) {
      const vulns = this._checkDependencyVulnerabilities(dep, ecosystem);
      if (vulns.length > 0) {
        check.vulnerableDependencies.push({
          name: dep.name,
          version: dep.version,
          vulnerabilities: vulns
        });

        for (const vuln of vulns) {
          check.summary.vulnerable++;
          check.summary[vuln.severity]++;
        }
      }
    }

    // Generate recommendations
    check.recommendations = this._generateDependencyRecommendations(check);

    return check;
  }

  /**
   * Analyze secrets in code
   */
  analyzeSecrets(content, strictMode = true) {
    const analysis = {
      timestamp: new Date().toISOString(),
      strictMode,
      secrets: [],
      summary: {
        total: 0,
        highConfidence: 0,
        mediumConfidence: 0,
        lowConfidence: 0
      }
    };

    // Scan for various types of secrets
    for (const [type, pattern] of Object.entries(this.secretPatterns)) {
      const matches = content.matchAll(pattern.regex);
      
      for (const match of matches) {
        const secret = {
          type,
          value: this._maskSecret(match[0]),
          line: this._findLineNumber(content, match.index),
          confidence: pattern.confidence,
          severity: pattern.severity,
          recommendation: pattern.recommendation
        };

        if (!strictMode || pattern.confidence === 'high') {
          analysis.secrets.push(secret);
          analysis.summary.total++;
          analysis.summary[`${pattern.confidence}Confidence`]++;
        }
      }
    }

    return analysis;
  }

  // Private helper methods

  _initVulnerabilityPatterns() {
    return {
      sqlInjection: [
        /\.query\s*\(\s*['"`][^'"`]*\$\{[^}]*\}[^'"`]*['"`]/g,
        /\.query\s*\(\s*['"`][^'"`]*\+[^)]+\)/g
      ],
      xss: [
        /innerHTML\s*=\s*[^;]+/g,
        /document\.write\s*\([^)]*\+[^)]*\)/g,
        /\$\([^)]+\)\.html\s*\([^)]*\+[^)]*\)/g
      ],
      commandInjection: [
        /exec\s*\(\s*['"`][^'"`]*\$\{[^}]*\}[^'"`]*['"`]/g,
        /spawn\s*\(\s*['"`][^'"`]*\$\{[^}]*\}[^'"`]*['"`]/g
      ]
    };
  }

  _initSecretPatterns() {
    return {
      'AWS Access Key': {
        regex: /AKIA[0-9A-Z]{16}/g,
        confidence: 'high',
        severity: 'critical',
        recommendation: 'Use AWS IAM roles or environment variables'
      },
      'API Key': {
        regex: /['"]?api[_-]?key['"]?\s*[:=]\s*['"]([a-zA-Z0-9_-]{20,})['"]/gi,
        confidence: 'medium',
        severity: 'high',
        recommendation: 'Store API keys in environment variables'
      },
      'Private Key': {
        regex: /-----BEGIN (RSA |EC |DSA |OPENSSH )?PRIVATE KEY-----/g,
        confidence: 'high',
        severity: 'critical',
        recommendation: 'Never commit private keys to source control'
      },
      'Password': {
        regex: /['"]?password['"]?\s*[:=]\s*['"]([^'"]{8,})['"]/gi,
        confidence: 'medium',
        severity: 'high',
        recommendation: 'Use environment variables and secret management'
      },
      'JWT Token': {
        regex: /eyJ[a-zA-Z0-9_-]+\.eyJ[a-zA-Z0-9_-]+\.[a-zA-Z0-9_-]+/g,
        confidence: 'high',
        severity: 'high',
        recommendation: 'Never commit tokens to source control'
      },
      'Database URL': {
        regex: /(postgres|mysql|mongodb):\/\/[^:]+:[^@]+@[^\s'"]+/gi,
        confidence: 'high',
        severity: 'critical',
        recommendation: 'Use environment variables for database credentials'
      }
    };
  }

  _detectSQLInjection(code) {
    const vulnerabilities = [];
    
    for (const pattern of this.vulnerabilityPatterns.sqlInjection) {
      const matches = code.matchAll(pattern);
      for (const match of matches) {
        vulnerabilities.push({
          type: 'SQL Injection',
          severity: 'critical',
          line: this._findLineNumber(code, match.index),
          code: match[0],
          description: 'Potential SQL injection vulnerability detected',
          remediation: 'Use parameterized queries or prepared statements',
          cwe: 'CWE-89',
          owasp: 'A03:2021 - Injection'
        });
      }
    }

    return vulnerabilities;
  }

  _detectXSS(code) {
    const vulnerabilities = [];
    
    for (const pattern of this.vulnerabilityPatterns.xss) {
      const matches = code.matchAll(pattern);
      for (const match of matches) {
        vulnerabilities.push({
          type: 'Cross-Site Scripting (XSS)',
          severity: 'high',
          line: this._findLineNumber(code, match.index),
          code: match[0],
          description: 'Potential XSS vulnerability detected',
          remediation: 'Sanitize user input and use safe DOM manipulation methods',
          cwe: 'CWE-79',
          owasp: 'A03:2021 - Injection'
        });
      }
    }

    return vulnerabilities;
  }

  _detectInsecureCrypto(code) {
    const vulnerabilities = [];
    
    if (/md5|sha1(?!\d)/gi.test(code)) {
      vulnerabilities.push({
        type: 'Weak Cryptographic Hash',
        severity: 'medium',
        line: this._findLineNumber(code, code.search(/md5|sha1(?!\d)/gi)),
        code: 'MD5/SHA1 usage',
        description: 'Use of weak cryptographic hash function',
        remediation: 'Use SHA-256 or stronger hash functions',
        cwe: 'CWE-327',
        owasp: 'A02:2021 - Cryptographic Failures'
      });
    }

    return vulnerabilities;
  }

  _detectCommandInjection(code) {
    const vulnerabilities = [];
    
    for (const pattern of this.vulnerabilityPatterns.commandInjection) {
      const matches = code.matchAll(pattern);
      for (const match of matches) {
        vulnerabilities.push({
          type: 'Command Injection',
          severity: 'critical',
          line: this._findLineNumber(code, match.index),
          code: match[0],
          description: 'Potential command injection vulnerability',
          remediation: 'Validate and sanitize input, use safe APIs',
          cwe: 'CWE-78',
          owasp: 'A03:2021 - Injection'
        });
      }
    }

    return vulnerabilities;
  }

  _detectPathTraversal(code) {
    const vulnerabilities = [];
    const pattern = /readFile\s*\(\s*[^)]*\+[^)]+\)|readFileSync\s*\(\s*[^)]*\+[^)]+\)/g;
    
    const matches = code.matchAll(pattern);
    for (const match of matches) {
      vulnerabilities.push({
        type: 'Path Traversal',
        severity: 'high',
        line: this._findLineNumber(code, match.index),
        code: match[0],
        description: 'Potential path traversal vulnerability',
        remediation: 'Validate file paths and use path.resolve()',
        cwe: 'CWE-22',
        owasp: 'A01:2021 - Broken Access Control'
      });
    }

    return vulnerabilities;
  }

  _detectInsecureDeserialization(code) {
    const vulnerabilities = [];
    
    if (/eval\s*\(|Function\s*\(|new\s+Function\s*\(/g.test(code)) {
      vulnerabilities.push({
        type: 'Insecure Deserialization',
        severity: 'critical',
        line: this._findLineNumber(code, code.search(/eval\s*\(/)),
        code: 'eval() or Function() usage',
        description: 'Use of dangerous deserialization methods',
        remediation: 'Use JSON.parse() or safe alternatives',
        cwe: 'CWE-502',
        owasp: 'A08:2021 - Software and Data Integrity Failures'
      });
    }

    return vulnerabilities;
  }

  _checkDependencyVulnerabilities(dep, _ecosystem) {
    // Simulated vulnerability database
    const knownVulnerabilities = {
      'lodash': {
        '<4.17.21': [
          {
            id: 'CVE-2021-23337',
            severity: 'high',
            description: 'Command injection vulnerability',
            fixedIn: '4.17.21'
          }
        ]
      },
      'express': {
        '<4.17.3': [
          {
            id: 'CVE-2022-24999',
            severity: 'medium',
            description: 'Open redirect vulnerability',
            fixedIn: '4.17.3'
          }
        ]
      }
    };

    const vulns = [];
    if (knownVulnerabilities[dep.name]) {
      for (const [versionRange, vulnList] of Object.entries(knownVulnerabilities[dep.name])) {
        // Simplified version comparison
        if (this._versionAffected(dep.version, versionRange)) {
          vulns.push(...vulnList);
        }
      }
    }

    return vulns;
  }

  _versionAffected(version, range) {
    // Simplified version comparison (in production, use semver library)
    if (range.startsWith('<')) {
      const targetVersion = range.substring(1);
      return version < targetVersion;
    }
    return false;
  }

  _generateDependencyRecommendations(check) {
    const recommendations = [];

    if (check.summary.critical > 0) {
      recommendations.push({
        priority: 'critical',
        action: 'Update dependencies with critical vulnerabilities immediately',
        packages: check.vulnerableDependencies
          .filter(d => d.vulnerabilities.some(v => v.severity === 'critical'))
          .map(d => d.name)
      });
    }

    if (check.summary.vulnerable > check.totalDependencies * 0.2) {
      recommendations.push({
        priority: 'high',
        action: 'Review and update dependency management strategy',
        impact: 'More than 20% of dependencies have known vulnerabilities'
      });
    }

    return recommendations;
  }

  _maskSecret(secret) {
    if (secret.length <= 8) {
return '***';
}
    return secret.substring(0, 4) + '***' + secret.substring(secret.length - 4);
  }

  _findLineNumber(content, index) {
    return content.substring(0, index).split('\n').length;
  }
}

/**
 * Get tool definitions for the security scanner
 */
function getToolDefinitions() {
  return [
    {
      name: 'scan-vulnerabilities',
      description: 'Scan code for security vulnerabilities',
      inputSchema: {
        type: 'object',
        properties: {
          code: { type: 'string', description: 'Code to scan' },
          language: { type: 'string', default: 'javascript' },
          severity: {
            type: 'string',
            enum: ['critical', 'high', 'medium', 'low', 'all'],
            default: 'all'
          }
        },
        required: ['code']
      }
    },
    {
      name: 'check-dependencies',
      description: 'Check dependencies for known vulnerabilities',
      inputSchema: {
        type: 'object',
        properties: {
          dependencies: {
            type: ['object', 'array'],
            description: 'Dependencies to check'
          },
          ecosystem: { type: 'string', default: 'npm' }
        },
        required: ['dependencies']
      }
    },
    {
      name: 'analyze-secrets',
      description: 'Analyze content for exposed secrets and credentials',
      inputSchema: {
        type: 'object',
        properties: {
          content: { type: 'string', description: 'Content to analyze' },
          strictMode: { type: 'boolean', default: true }
        },
        required: ['content']
      }
    }
  ];
}

/**
 * Handle tool execution requests
 */
function handleToolCall(scanner, name, args) {
  switch (name) {
    case 'scan-vulnerabilities': {
      const validated = ScanVulnerabilitiesSchema.parse(args);
      const result = scanner.scanVulnerabilities(
        validated.code,
        validated.language,
        validated.severity
      );
      return {
        content: [{ type: 'text', text: JSON.stringify(result, null, 2) }]
      };
    }

    case 'check-dependencies': {
      const validated = CheckDependenciesSchema.parse(args);
      const result = scanner.checkDependencies(
        validated.dependencies,
        validated.ecosystem
      );
      return {
        content: [{ type: 'text', text: JSON.stringify(result, null, 2) }]
      };
    }

    case 'analyze-secrets': {
      const validated = AnalyzeSecretsSchema.parse(args);
      const result = scanner.analyzeSecrets(
        validated.content,
        validated.strictMode
      );
      return {
        content: [{ type: 'text', text: JSON.stringify(result, null, 2) }]
      };
    }

    default:
      throw new McpError(ErrorCode.MethodNotFound, `Unknown tool: ${name}`);
  }
}

/**
 * Initialize and start the MCP server
 */
async function main() {
  const scanner = new SecurityScanner();
  const server = new Server(
    {
      name: 'security-scanner',
      version: '1.0.0',
    },
    {
      capabilities: {
        tools: {},
      },
    }
  );

  server.setRequestHandler(ListToolsRequestSchema, () => ({
    tools: getToolDefinitions()
  }));

  server.setRequestHandler(CallToolRequestSchema, (request) => {
    const { name, arguments: args } = request.params;

    try {
      return handleToolCall(scanner, name, args);
    } catch (error) {
      if (error instanceof z.ZodError) {
        throw new McpError(
          ErrorCode.InvalidParams,
          `Invalid parameters: ${error.errors.map(e => e.message).join(', ')}`
        );
      }
      if (error instanceof McpError) {
        throw error;
      }
      throw new McpError(
        ErrorCode.InternalError,
        error instanceof Error ? error.message : String(error)
      );
    }
  });

  const transport = new StdioServerTransport();
  await server.connect(transport);

  console.error('Security Scanner MCP Server started successfully');
}

main().catch((error) => {
  console.error('Fatal error:', error);
  process.exit(1);
});
